{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Part 1: Data Parsing\n",
    "\n",
    "In this part of the tutorials, you will be guided through the different parsing steps needed to ensure high-quality consistent data formats. <br>\n",
    "<br>\n",
    "This includes: \n",
    "- Loading the data\n",
    "- Column selection\n",
    "- Remove entries with missing data\n",
    "- Remove CDR3β sequences containing non-amino acid characters\n",
    "- Remove CDR3β sequences not starting/ending with the conserved cysteine (C) and end with a phenylalanine (F) residue\n",
    "- Merging and excluding duplicate sequences from the data\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import packages\n",
    "\n",
    "import warnings\n",
    "warnings.filterwarnings('ignore')\n",
    "\n",
    "import pandas as pd\n",
    "\n",
    "from pyteomics import parser"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load the data\n",
    "\n",
    "# Define your current working directory\n",
    "cwd = '/your/working/directory'\n",
    "\n",
    "# Read in the correct data files\n",
    "Patient_1_pre = pd.read_csv(f'{cwd}/Data/HN005_TUMOR_PRE.tsv', sep='\\t')\n",
    "Patient_1_post = pd.read_csv(f'{cwd}/Data/HN005_TUMOR_POST.tsv', sep='\\t')\n",
    "\n",
    "Patient_2_pre = pd.read_csv(f'{cwd}/Data/HN006_TUMOR_PRE.tsv', sep='\\t')\n",
    "Patient_2_post = pd.read_csv(f'{cwd}/Data/HN006_TUMOR_POST.tsv', sep='\\t')\n",
    "\n",
    "Patient_3_pre = pd.read_csv(f'{cwd}/Data/HN009_TUMOR_PRE.tsv', sep='\\t')\n",
    "Patient_3_post = pd.read_csv(f'{cwd}/Data/HN009_TUMOR_POST.tsv', sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>nucleotide</th>\n",
       "      <th>aminoAcid</th>\n",
       "      <th>count (templates/reads)</th>\n",
       "      <th>frequencyCount (%)</th>\n",
       "      <th>cdr3Length</th>\n",
       "      <th>vMaxResolved</th>\n",
       "      <th>vFamilyName</th>\n",
       "      <th>vGeneName</th>\n",
       "      <th>vGeneAllele</th>\n",
       "      <th>vFamilyTies</th>\n",
       "      <th>...</th>\n",
       "      <th>jOrphon</th>\n",
       "      <th>vFunction</th>\n",
       "      <th>dFunction</th>\n",
       "      <th>jFunction</th>\n",
       "      <th>fractionNucleated</th>\n",
       "      <th>vAlignLength</th>\n",
       "      <th>vAlignSubstitutionCount</th>\n",
       "      <th>vAlignSubstitutionIndexes</th>\n",
       "      <th>vAlignSubstitutionGeneThreePrimeIndexes</th>\n",
       "      <th>vSeqWithMutations</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AGTGCCCATCCTGAAGACAGCAGCTTCTACATCTGCAGTGCTAATC...</td>\n",
       "      <td>CSANR*D*RGYNEQFF</td>\n",
       "      <td>358</td>\n",
       "      <td>0.011601</td>\n",
       "      <td>48</td>\n",
       "      <td>TCRBV20</td>\n",
       "      <td>TCRBV20</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AGTGCCCATCCTGAAGACAGCAGCTTCTACATCTGCAGTGCTAGTC...</td>\n",
       "      <td>CSASRADRTPIYGYTF</td>\n",
       "      <td>283</td>\n",
       "      <td>0.009171</td>\n",
       "      <td>48</td>\n",
       "      <td>TCRBV20</td>\n",
       "      <td>TCRBV20</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CACGCCCTGCAGCCAGAAGACTCAGCCCTGTATCTCTGCGCCAGCA...</td>\n",
       "      <td>CASSQARGINQPQHF</td>\n",
       "      <td>325</td>\n",
       "      <td>0.010532</td>\n",
       "      <td>45</td>\n",
       "      <td>TCRBV04-01*01</td>\n",
       "      <td>TCRBV04</td>\n",
       "      <td>TCRBV04-01</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CCAGGCCTGGGGGACGCAGCCATGTACCTGTGTGCCACCAGCAGAG...</td>\n",
       "      <td>CATSRESPGQGIDEQFF</td>\n",
       "      <td>228</td>\n",
       "      <td>0.007388</td>\n",
       "      <td>51</td>\n",
       "      <td>TCRBV15-01*01</td>\n",
       "      <td>TCRBV15</td>\n",
       "      <td>TCRBV15-01</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AGCTCCTTCTCAGTGACTCTGGCTTCTATCTCTGTGCCTGGAGCTG...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>148</td>\n",
       "      <td>0.004796</td>\n",
       "      <td>49</td>\n",
       "      <td>TCRBV30-01*01</td>\n",
       "      <td>TCRBV30</td>\n",
       "      <td>TCRBV30-01</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14560</th>\n",
       "      <td>CCTGAGCTCTCTGAGCTGGGGGACTCAGCTTTGTATTTCTGTGCCA...</td>\n",
       "      <td>CASSVGQANTEAFF</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000032</td>\n",
       "      <td>42</td>\n",
       "      <td>TCRBV09-01</td>\n",
       "      <td>TCRBV09</td>\n",
       "      <td>TCRBV09-01</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14561</th>\n",
       "      <td>GAGTTCTAAGAAGCTCCTTCTCAGTGACTCTGGCTTCTATCTCTGT...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000032</td>\n",
       "      <td>38</td>\n",
       "      <td>TCRBV30-01*01</td>\n",
       "      <td>TCRBV30</td>\n",
       "      <td>TCRBV30-01</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14562</th>\n",
       "      <td>CTTGTCCACTCTGACAGTGACCAGTGCCATCCTGAAGACAGCAGCT...</td>\n",
       "      <td>CSARDEAFF</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000032</td>\n",
       "      <td>27</td>\n",
       "      <td>TCRBV20</td>\n",
       "      <td>TCRBV20</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14563</th>\n",
       "      <td>GAGCACCTTGGAGCTGGGGGACTCGGCCCTTTATCTTTGCGCCAGC...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000032</td>\n",
       "      <td>44</td>\n",
       "      <td>TCRBV05-01*01</td>\n",
       "      <td>TCRBV05</td>\n",
       "      <td>TCRBV05-01</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14564</th>\n",
       "      <td>CCTCAGAACCCAGGGACTCAGCTGTGTACTTCTGTGCCAGCAGTTT...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000032</td>\n",
       "      <td>49</td>\n",
       "      <td>TCRBV12</td>\n",
       "      <td>TCRBV12</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>14565 rows × 52 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                              nucleotide          aminoAcid  \\\n",
       "0      AGTGCCCATCCTGAAGACAGCAGCTTCTACATCTGCAGTGCTAATC...   CSANR*D*RGYNEQFF   \n",
       "1      AGTGCCCATCCTGAAGACAGCAGCTTCTACATCTGCAGTGCTAGTC...   CSASRADRTPIYGYTF   \n",
       "2      CACGCCCTGCAGCCAGAAGACTCAGCCCTGTATCTCTGCGCCAGCA...    CASSQARGINQPQHF   \n",
       "3      CCAGGCCTGGGGGACGCAGCCATGTACCTGTGTGCCACCAGCAGAG...  CATSRESPGQGIDEQFF   \n",
       "4      AGCTCCTTCTCAGTGACTCTGGCTTCTATCTCTGTGCCTGGAGCTG...                NaN   \n",
       "...                                                  ...                ...   \n",
       "14560  CCTGAGCTCTCTGAGCTGGGGGACTCAGCTTTGTATTTCTGTGCCA...     CASSVGQANTEAFF   \n",
       "14561  GAGTTCTAAGAAGCTCCTTCTCAGTGACTCTGGCTTCTATCTCTGT...                NaN   \n",
       "14562  CTTGTCCACTCTGACAGTGACCAGTGCCATCCTGAAGACAGCAGCT...          CSARDEAFF   \n",
       "14563  GAGCACCTTGGAGCTGGGGGACTCGGCCCTTTATCTTTGCGCCAGC...                NaN   \n",
       "14564  CCTCAGAACCCAGGGACTCAGCTGTGTACTTCTGTGCCAGCAGTTT...                NaN   \n",
       "\n",
       "       count (templates/reads)  frequencyCount (%)  cdr3Length   vMaxResolved  \\\n",
       "0                          358            0.011601          48        TCRBV20   \n",
       "1                          283            0.009171          48        TCRBV20   \n",
       "2                          325            0.010532          45  TCRBV04-01*01   \n",
       "3                          228            0.007388          51  TCRBV15-01*01   \n",
       "4                          148            0.004796          49  TCRBV30-01*01   \n",
       "...                        ...                 ...         ...            ...   \n",
       "14560                        1            0.000032          42     TCRBV09-01   \n",
       "14561                        1            0.000032          38  TCRBV30-01*01   \n",
       "14562                        1            0.000032          27        TCRBV20   \n",
       "14563                        1            0.000032          44  TCRBV05-01*01   \n",
       "14564                        1            0.000032          49        TCRBV12   \n",
       "\n",
       "      vFamilyName   vGeneName  vGeneAllele vFamilyTies  ... jOrphon vFunction  \\\n",
       "0         TCRBV20         NaN          NaN         NaN  ...     NaN       NaN   \n",
       "1         TCRBV20         NaN          NaN         NaN  ...     NaN       NaN   \n",
       "2         TCRBV04  TCRBV04-01          1.0         NaN  ...     NaN       NaN   \n",
       "3         TCRBV15  TCRBV15-01          1.0         NaN  ...     NaN       NaN   \n",
       "4         TCRBV30  TCRBV30-01          1.0         NaN  ...     NaN       NaN   \n",
       "...           ...         ...          ...         ...  ...     ...       ...   \n",
       "14560     TCRBV09  TCRBV09-01          NaN         NaN  ...     NaN       NaN   \n",
       "14561     TCRBV30  TCRBV30-01          1.0         NaN  ...     NaN       NaN   \n",
       "14562     TCRBV20         NaN          NaN         NaN  ...     NaN       NaN   \n",
       "14563     TCRBV05  TCRBV05-01          1.0         NaN  ...     NaN       NaN   \n",
       "14564     TCRBV12         NaN          NaN         NaN  ...     NaN       NaN   \n",
       "\n",
       "      dFunction jFunction fractionNucleated  vAlignLength  \\\n",
       "0           NaN       NaN               NaN           NaN   \n",
       "1           NaN       NaN               NaN           NaN   \n",
       "2           NaN       NaN               NaN           NaN   \n",
       "3           NaN       NaN               NaN           NaN   \n",
       "4           NaN       NaN               NaN           NaN   \n",
       "...         ...       ...               ...           ...   \n",
       "14560       NaN       NaN               NaN           NaN   \n",
       "14561       NaN       NaN               NaN           NaN   \n",
       "14562       NaN       NaN               NaN           NaN   \n",
       "14563       NaN       NaN               NaN           NaN   \n",
       "14564       NaN       NaN               NaN           NaN   \n",
       "\n",
       "      vAlignSubstitutionCount vAlignSubstitutionIndexes  \\\n",
       "0                         NaN                       NaN   \n",
       "1                         NaN                       NaN   \n",
       "2                         NaN                       NaN   \n",
       "3                         NaN                       NaN   \n",
       "4                         NaN                       NaN   \n",
       "...                       ...                       ...   \n",
       "14560                     NaN                       NaN   \n",
       "14561                     NaN                       NaN   \n",
       "14562                     NaN                       NaN   \n",
       "14563                     NaN                       NaN   \n",
       "14564                     NaN                       NaN   \n",
       "\n",
       "      vAlignSubstitutionGeneThreePrimeIndexes vSeqWithMutations  \n",
       "0                                         NaN               NaN  \n",
       "1                                         NaN               NaN  \n",
       "2                                         NaN               NaN  \n",
       "3                                         NaN               NaN  \n",
       "4                                         NaN               NaN  \n",
       "...                                       ...               ...  \n",
       "14560                                     NaN               NaN  \n",
       "14561                                     NaN               NaN  \n",
       "14562                                     NaN               NaN  \n",
       "14563                                     NaN               NaN  \n",
       "14564                                     NaN               NaN  \n",
       "\n",
       "[14565 rows x 52 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Patient_3_pre"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define parsing functions\n",
    "\n",
    "def get_columns(data):\n",
    "\n",
    "    # Extract informative columns from the large dataframes\n",
    "    result = data[['aminoAcid','vFamilyName', 'jFamilyName',  'count (templates/reads)', 'frequencyCount (%)', 'sequenceStatus']]\n",
    "\n",
    "    # Include only the productive TCRs as described by Immuneaccess\n",
    "    result = result.loc[result['sequenceStatus'] == 'In']\n",
    "\n",
    "    # Change column names and add a column describing the full sequence (V- and J-family included)\n",
    "    result.rename(columns={'aminoAcid': 'junction_aa', 'count (templates/reads)': 'Clone_count', 'frequencyCount (%)': 'Clone_frequency', 'vFamilyName': 'v_call', 'jFamilyName': 'j_call'}, inplace=True)\n",
    "    result['Full_CDR3'] = result['junction_aa'] + '_' + result['v_call'] + '_' + result['j_call']\n",
    "    result = result.dropna()\n",
    "\n",
    "    return result\n",
    "\n",
    "\n",
    "# Define all known amino acids\n",
    "def is_amino_acid(cdr3):\n",
    "    return all(aa in parser.std_amino_acids for aa in cdr3)\n",
    "\n",
    "\n",
    "def parse_data(dataframe):\n",
    "\n",
    "    # Select only the sequences with known amino acids\n",
    "    Step1 = dataframe[dataframe['junction_aa'].apply(is_amino_acid)]\n",
    "\n",
    "    # Make sure each sequence starts with C and ends with F\n",
    "    start_c = Step1['junction_aa'].str.startswith('C', na=False)\n",
    "    end_f = Step1['junction_aa'].str.endswith('F', na=False)\n",
    "    parse_result = Step1[start_c & end_f].reset_index(drop=True)\n",
    "\n",
    "    # Sum the clone counts for duplicate TCRs and merge into one entry\n",
    "    parse_result['Total_count'] = parse_result.groupby(['junction_aa','v_call','j_call'])['Clone_count'].transform('sum')\n",
    "    parse_result['Total_frequency'] = parse_result.groupby(['junction_aa','v_call','j_call'])['Clone_frequency'].transform('sum')\n",
    "    parse_result = parse_result[['junction_aa', 'v_call', 'j_call', 'Full_CDR3', 'Total_count', 'Total_frequency']]\n",
    "    parse_result.drop_duplicates(subset=['junction_aa', 'v_call', 'j_call'], inplace=True)\n",
    "\n",
    "    return parse_result\n",
    "\n",
    "\n",
    "# Apply column extraction and data parsing on dataframe\n",
    "def full_parsing(data, patient):\n",
    "    Columns = get_columns(data)\n",
    "    Parsing = parse_data(Columns)\n",
    "    print(f'Parsing: {patient}')\n",
    "    print(f'The number of TCRs before parsing is: {len(Columns)}')\n",
    "    print(f'The number of TCRs after parsing is: {len(Parsing)}')\n",
    "    print(f'The number of TCRs removed is : {len(Columns) - len(Parsing)} \\n')\n",
    "\n",
    "    return Parsing\n",
    "\n",
    "\n",
    "# Get total number of TCRs per patient\n",
    "def get_patient_total(data_pre, data_post, Patient):\n",
    "    data_pre = pd.DataFrame(data_pre['Full_CDR3'])\n",
    "    data_post = pd.DataFrame(data_post['Full_CDR3'])\n",
    "    merged = pd.concat([data_pre,data_post]).drop_duplicates().reset_index(drop=True)\n",
    "    print('The total number of unique sequences for', Patient, 'is:', len(merged))\n",
    "\n",
    "    return merged"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Apply the parsing function defined above on the TCR data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Parsing: Patient_1_pre\n",
      "The number of TCRs before parsing is: 15881\n",
      "The number of TCRs after parsing is: 15074\n",
      "The number of TCRs removed is : 807 \n",
      "\n",
      "Parsing: Patient_1_post\n",
      "The number of TCRs before parsing is: 2929\n",
      "The number of TCRs after parsing is: 2720\n",
      "The number of TCRs removed is : 209 \n",
      "\n",
      "Parsing: Patient_2_pre\n",
      "The number of TCRs before parsing is: 19456\n",
      "The number of TCRs after parsing is: 17976\n",
      "The number of TCRs removed is : 1480 \n",
      "\n",
      "Parsing: Patient_2_post\n",
      "The number of TCRs before parsing is: 43222\n",
      "The number of TCRs after parsing is: 39808\n",
      "The number of TCRs removed is : 3414 \n",
      "\n",
      "Parsing: Patient_3_pre\n",
      "The number of TCRs before parsing is: 11836\n",
      "The number of TCRs after parsing is: 11584\n",
      "The number of TCRs removed is : 252 \n",
      "\n",
      "Parsing: Patient_3_post\n",
      "The number of TCRs before parsing is: 48045\n",
      "The number of TCRs after parsing is: 46629\n",
      "The number of TCRs removed is : 1416 \n",
      "\n",
      "The total number of unique sequences for Patient 1 is: 16228\n",
      "The total number of unique sequences for Patient 2 is: 49769\n",
      "The total number of unique sequences for Patient 3 is: 52905\n"
     ]
    }
   ],
   "source": [
    "# Parse data\n",
    "\n",
    "# Execute parsing steps for each dataset\n",
    "Data = [Patient_1_pre, Patient_1_post, Patient_2_pre, Patient_2_post, Patient_3_pre, Patient_3_post]\n",
    "Data_names = ['Patient_1_pre', 'Patient_1_post', 'Patient_2_pre', 'Patient_2_post', 'Patient_3_pre', 'Patient_3_post']\n",
    "Parsed_frames = []\n",
    "\n",
    "for i,j in zip(Data,Data_names):\n",
    "    res = full_parsing(i,j)\n",
    "    Parsed_frames.append(res)\n",
    "\n",
    "P1_pre, P1_post, P2_pre, P2_post, P3_pre, P3_post = [pd.DataFrame(j) for j in Parsed_frames]\n",
    "\n",
    "# Get total number of unique sequences for each patient\n",
    "Patient1_total = get_patient_total(P1_pre, P1_post, 'Patient 1')\n",
    "Patient2_total = get_patient_total(P2_pre, P2_post, 'Patient 2')\n",
    "Patient3_total = get_patient_total(P3_pre, P3_post, 'Patient 3')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>junction_aa</th>\n",
       "      <th>v_call</th>\n",
       "      <th>j_call</th>\n",
       "      <th>Full_CDR3</th>\n",
       "      <th>Total_count</th>\n",
       "      <th>Total_frequency</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CSARDTVHYNEQFF</td>\n",
       "      <td>TCRBV20</td>\n",
       "      <td>TCRBJ02</td>\n",
       "      <td>CSARDTVHYNEQFF_TCRBV20_TCRBJ02</td>\n",
       "      <td>1769</td>\n",
       "      <td>0.020675</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CASRSRESNTIYF</td>\n",
       "      <td>TCRBV06</td>\n",
       "      <td>TCRBJ01</td>\n",
       "      <td>CASRSRESNTIYF_TCRBV06_TCRBJ01</td>\n",
       "      <td>1975</td>\n",
       "      <td>0.023082</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CASSPRDWTYEQYF</td>\n",
       "      <td>TCRBV07</td>\n",
       "      <td>TCRBJ02</td>\n",
       "      <td>CASSPRDWTYEQYF_TCRBV07_TCRBJ02</td>\n",
       "      <td>1253</td>\n",
       "      <td>0.014644</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CASRTSGSTDTQYF</td>\n",
       "      <td>TCRBV06</td>\n",
       "      <td>TCRBJ02</td>\n",
       "      <td>CASRTSGSTDTQYF_TCRBV06_TCRBJ02</td>\n",
       "      <td>1125</td>\n",
       "      <td>0.013148</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CASTLAGPTPVEQYF</td>\n",
       "      <td>TCRBV03</td>\n",
       "      <td>TCRBJ02</td>\n",
       "      <td>CASTLAGPTPVEQYF_TCRBV03_TCRBJ02</td>\n",
       "      <td>839</td>\n",
       "      <td>0.009806</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15849</th>\n",
       "      <td>CASSGLRGTEAFF</td>\n",
       "      <td>TCRBV02</td>\n",
       "      <td>TCRBJ01</td>\n",
       "      <td>CASSGLRGTEAFF_TCRBV02_TCRBJ01</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000012</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15851</th>\n",
       "      <td>CAISEVLDSTEAFF</td>\n",
       "      <td>TCRBV10</td>\n",
       "      <td>TCRBJ01</td>\n",
       "      <td>CAISEVLDSTEAFF_TCRBV10_TCRBJ01</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000012</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15854</th>\n",
       "      <td>CASSLDAVNTEAFF</td>\n",
       "      <td>TCRBV07</td>\n",
       "      <td>TCRBJ01</td>\n",
       "      <td>CASSLDAVNTEAFF_TCRBV07_TCRBJ01</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000012</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15855</th>\n",
       "      <td>CASTEGALGNTEAFF</td>\n",
       "      <td>TCRBV05</td>\n",
       "      <td>TCRBJ01</td>\n",
       "      <td>CASTEGALGNTEAFF_TCRBV05_TCRBJ01</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000012</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15858</th>\n",
       "      <td>CASSQNREAFF</td>\n",
       "      <td>TCRBV11</td>\n",
       "      <td>TCRBJ01</td>\n",
       "      <td>CASSQNREAFF_TCRBV11_TCRBJ01</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000012</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>15074 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           junction_aa   v_call   j_call                        Full_CDR3  \\\n",
       "0       CSARDTVHYNEQFF  TCRBV20  TCRBJ02   CSARDTVHYNEQFF_TCRBV20_TCRBJ02   \n",
       "1        CASRSRESNTIYF  TCRBV06  TCRBJ01    CASRSRESNTIYF_TCRBV06_TCRBJ01   \n",
       "2       CASSPRDWTYEQYF  TCRBV07  TCRBJ02   CASSPRDWTYEQYF_TCRBV07_TCRBJ02   \n",
       "3       CASRTSGSTDTQYF  TCRBV06  TCRBJ02   CASRTSGSTDTQYF_TCRBV06_TCRBJ02   \n",
       "4      CASTLAGPTPVEQYF  TCRBV03  TCRBJ02  CASTLAGPTPVEQYF_TCRBV03_TCRBJ02   \n",
       "...                ...      ...      ...                              ...   \n",
       "15849    CASSGLRGTEAFF  TCRBV02  TCRBJ01    CASSGLRGTEAFF_TCRBV02_TCRBJ01   \n",
       "15851   CAISEVLDSTEAFF  TCRBV10  TCRBJ01   CAISEVLDSTEAFF_TCRBV10_TCRBJ01   \n",
       "15854   CASSLDAVNTEAFF  TCRBV07  TCRBJ01   CASSLDAVNTEAFF_TCRBV07_TCRBJ01   \n",
       "15855  CASTEGALGNTEAFF  TCRBV05  TCRBJ01  CASTEGALGNTEAFF_TCRBV05_TCRBJ01   \n",
       "15858      CASSQNREAFF  TCRBV11  TCRBJ01      CASSQNREAFF_TCRBV11_TCRBJ01   \n",
       "\n",
       "       Total_count  Total_frequency  \n",
       "0             1769         0.020675  \n",
       "1             1975         0.023082  \n",
       "2             1253         0.014644  \n",
       "3             1125         0.013148  \n",
       "4              839         0.009806  \n",
       "...            ...              ...  \n",
       "15849            1         0.000012  \n",
       "15851            1         0.000012  \n",
       "15854            1         0.000012  \n",
       "15855            1         0.000012  \n",
       "15858            1         0.000012  \n",
       "\n",
       "[15074 rows x 6 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Example of resulting dataframe\n",
    "\n",
    "P1_pre"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Export parsed data to file\n",
    "\n",
    "Dataframes = [P1_pre, P1_post, P2_pre, P2_post, P3_pre, P3_post, Patient1_total, Patient2_total, Patient3_total]\n",
    "Names=['P1_pre', 'P1_post', 'P2_pre', 'P2_post', 'P3_pre', 'P3_post', 'Patient1_total', 'Patient2_total', 'Patient3_total']\n",
    "\n",
    "for i,j in zip(Dataframes, Names):\n",
    "    i.to_csv(f'{cwd}/Processed_data/{j}_data.tsv', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Export part of parsed data to files for clustering\n",
    "\n",
    "Dataframes = [P1_pre, P1_post, P2_pre, P2_post, P3_pre, P3_post]\n",
    "Names=['P1_pre', 'P1_post', 'P2_pre', 'P2_post', 'P3_pre', 'P3_post']\n",
    "\n",
    "for i,j in zip(Dataframes, Names):\n",
    "    i = i[['junction_aa', 'v_call', 'j_call']]\n",
    "    i.to_csv(f'{cwd}/Processed_data/{j}_cluster_data.tsv', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "interpreter": {
   "hash": "4d59cba34f833ae973e4f300577f70c1cfbbaccacd8edc2839287c82209f74ba"
  },
  "kernelspec": {
   "display_name": "Python 3.8.15 64-bit ('tcr': conda)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.15"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
